pkg/endpoint/bpf.go

```diff
@@ -242,34 +251,153 @@ func (e *Endpoint) addNewRedirectsFromMap(m policy.L4PolicyMap, desiredRedirects
+		// redirect for that port.
+		e.proxyStatisticsMutex.Lock()
+		proxyStats := e.getProxyStatisticsLocked(proxyID, string(visMeta.Parser), visMeta.Port, visMeta.Ingress)
+		proxyStats.AllocatedProxyPort = int64(redirectPort)
+		e.proxyStatisticsMutex.Unlock()
+
+		updatedStats = append(updatedStats, proxyStats)
+
+		newKey := policy.Key{
+			DestPort:         visMeta.Port,
+			Nexthdr:          uint8(visMeta.Proto),
+			TrafficDirection: direction.Uint8(),
+		}
+
+		e.desiredPolicy.PolicyMapState[newKey] = policy.MapStateEntry{
+			ProxyPort: redirectPort,
+		}
+
+		insertedDesiredMapState[newKey] = struct{}{}
+	}
+
+	revertStack.Push(func() error {
+		// Restore the proxy stats.
+		e.proxyStatisticsMutex.Lock()
+		for _, stats := range updatedStats {
+			stats.AllocatedProxyPort = 0
+		}
+		e.proxyStatisticsMutex.Unlock()
+
+		// Restore the desired policy map state.
+		for key := range insertedDesiredMapState {
+			delete(e.desiredPolicy.PolicyMapState, key)
+		}
+		return nil
+	})
+
+	return nil, finalizeList.Finalize, revertStack.Revert
+}
+
 // addNewRedirects must be called while holding the endpoint lock for writing.
 // On success, returns nil; otherwise, returns an error indicating the problem
 // that occurred while adding an l7 redirect for the specified policy.
 // The returned map contains the exact set of IDs of proxy redirects that is
 // required to implement the given L4 policy.
 // Must be called with endpoint.Mutex held.
 func (e *Endpoint) addNewRedirects(m *policy.L4Policy, proxyWaitGroup *completion.WaitGroup) (desiredRedirects map[string]bool, err error, finalizeFunc revert.FinalizeFunc, revertFunc revert.RevertFunc) {
-	desiredRedirects = make(map[string]bool)
-	var finalizeList revert.FinalizeList
-	var revertStack revert.RevertStack
+	var (
+		finalizeList revert.FinalizeList
+		revertStack  revert.RevertStack
+		ff           revert.FinalizeFunc
+		rf           revert.RevertFunc
+	)
 
-	var ff revert.FinalizeFunc
-	var rf revert.RevertFunc
+	defer func() {
+		// In case updates partially succeeded, and subsequently failed,
+		// revert.
+		if err != nil {
+			revertStack.Revert()
+		}
+	}()
 
-	err, ff, rf = e.addNewRedirectsFromMap(m.Ingress, desiredRedirects, proxyWaitGroup)
-	if err != nil {
-		return desiredRedirects, fmt.Errorf("unable to allocate ingress redirects: %s", err), nil, nil
-	}
-	finalizeList.Append(ff)
-	revertStack.Push(rf)
+	desiredRedirects = make(map[string]bool)
 
-	err, ff, rf = e.addNewRedirectsFromMap(m.Egress, desiredRedirects, proxyWaitGroup)
-	if err != nil {
-		revertStack.Revert() // Ignore errors while reverting. This is best-effort.
-		return desiredRedirects, fmt.Errorf("unable to allocate egress redirects: %s", err), nil, nil
+	for dirLogStr, ingress := range map[string]bool{"ingress": true, "egress": false} {
+		if m != nil && m.HasRedirect() {
+			err, ff, rf = e.addNewRedirectsFromMap(ingress, desiredRedirects, proxyWaitGroup)
```

Comment:
Note: This change in passing down the policy means that `addNewRedirects()` effectively ignores the first parameter, the policy. Either `addNewRedirects()` should drop the policy argument, or this code should actually pass down some part of the policy that is specified in the first parameter to `addNewRedirects()`. (I'm guessing the former)

Reply:
Fixed.